Allow block-specific custom volume quadrature rules

rwcarlsen opened the pull request:
reopening of #14214.

rwcarlsen committed:
Allow per-block quadrature rules to be specified.

Create a new quadrature rule data-structure to consolidate several
others and use unique pointers instead of raw.

fixes #14055

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/548319/) on 17bec54 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15470/site)

This comment will be updated on new commits.

rwcarlsen committed:
add a test

rwcarlsen committed:
Allow custom bumping of elem block quadrature.

Also update/fix max qps computation to use the block-specific qrules.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/550895/) on b9f3df9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15470/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15470/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 305d0f125a9912de786ad362b5a787e67698a005`

rwcarlsen commented:
@YaqiWang - how is this?  Check out the `bumpVolumeQRuleOrder` function.

rwcarlsen committed:
add doc comments

lindsayad commented on framework/include/base/Assembly.h:
```diff
@@ -1394,6 +1416,17 @@ class Assembly
+   * Returns the current elem/volume quadrature rule.  The current subdomain
+   * (as set via setCurrentSubdomainID is used to determine the correct rule.
+   */
+  inline QBase * qruleElem(unsigned int dim) { return qrules(dim).vol.get(); }
```
If this is going to be public, then I’d like to see this return a const QBase * and have the method marked const.  Then if you want to be able to use this for setting purposes within Assembly, then you could add a private non-const method overload that returns a QBase *. Same for the face method

rwcarlsen replied:
I actually tried to make them const at first.  But after attaching a quadrature rule to an fe object - they get init'd - so any actually used quadrature rules (which is the purpose of these accessors) will need to be available as non-const.

lindsayad replied:
What about about adding in `Assembly`
```
void attachQRuleElem(FEBase & fe, unsigned int dim)
{
  fe->attach_quadrature_rule(qRuleElem(dim));
}
```
I just think that this non-const accessor could really be abused by users. Someone could just willy nilly set the qrule to something crazy.

lindsayad replied:
Or sorry, they couldn't change the Assembly qrule to a different object, but they could mess with the qrule through the `get_points,get_weights,etc.` APIs. It's probably unlikely but I get paranoid

rwcarlsen replied:
Yeah.  Too bad QBase doesn't have a virtual copy-constructor pattern clone func we can call to just copy it.  That would be the best solution I think.

rwcarlsen replied:
Okay - this should be addressed.

lindsayad reviewed (changes requested):
This looks really good to me. Just the one const concern

rwcarlsen committed:
fix nonunity build

rwcarlsen committed:
Make qrule accessors more safe.

We need to allow a non-const init operation that occurs when attaching
qrules to fe objects.  But other than that we want to only allow const
access.

lindsayad reviewed (dismissed):
Thanks @rwcarlsen for the change. I really like this

lindsayad commented:
I'll let @YaqiWang comment on whether this meets his desires

YaqiWang commented on framework/include/base/Assembly.h:
```diff
@@ -434,10 +438,29 @@ class Assembly
+   * orders passed in.  Any quadrature rules specified using this function
+   * override those created via in the non-block-specific/global createQRules
+   * function.
+   */
```
Need documenting the parameters. It is not so clear to me the difference of the three order parameters.

rwcarlsen replied:
Yeah - I was just being lazy here since the original createQRules function didn't say anything about them either.  But really - only the `order` parameter is unclear I think.  I'll push up a clarification.

YaqiWang replied:
yes.

YaqiWang commented on framework/include/base/Assembly.h:
```diff
@@ -434,10 +438,29 @@ class Assembly
+   * lower than or equal to the current volume/elem quadrature rule order,
+   * then nothing is done (i.e. this function is idempotent).
+   */
+  void bumpVolumeQRuleOrder(Order volume_order, SubdomainID block);
```
Why do we need two functions, this and the above to set orders?

rwcarlsen replied:
Because you don't know what the orders may have been set to.  And the above functions recreate *all* the quadrature rules.  If you just want to bump volume quadrature, you would have to know/re-specify the face and general orders again to be what they were - except you don't know what they were.

YaqiWang commented on test/tests/quadrature/order/block-order.i:
```diff
@@ -0,0 +1,57 @@
+
+[Outputs]
+  execute_on = 'timestep_end'
+  exodus = false
```
Need a test for the function `bumpVolumeQRuleOrder`. We can create a dummy material in `test/src/materials`, which calls that function.

rwcarlsen replied:
A test is a good idea (I pushed one up).  Yes - generally quadrature adjustments need to happen early (i.e. in constructors).

rwcarlsen committed:
add in-code quadrature bump test

rwcarlsen commented:
@lindsayad - could you approve this again - I pushed up an extra test.

vincentlaboure commented:
That should be really useful, thank you @rwcarlsen !

YaqiWang commented:
I am testing this with Griffin, but this seems not doing exactly what I want. Let us say we have two blocks 0 and 1, block 0 has quadrature order 3, block 1 has quadrature order 5. On the interface of the two blocks, I'd like MOOSE to choose the higher one, i.e. 5 always. But I guess currently the face quadrature order depends on the currently element. If on the interface, the current element is in block 0, MOOSE will use order 3 instead of order 5. Should we always compare the two orders and choose the larger one? I am asking for this because the face integration needs to be done with higher order due to the requirement from the high-order subdomain. Tag @rwcarlsen

rwcarlsen commented:
Currently - it doesn't do anything with face quadrature rules.  They global defaults are used - none of the per-block stuff factors into face quadrature rules at all.

rwcarlsen commented:
We punted on that.  Are you doing DG stuff or are these faces on boundaries (i.e. boundary id is defined).

YaqiWang commented:
We have both internal sides and boundary sides. The interface between block 0 and block 1 is not necessary a side set. We had some talks on this but I forgot what those ended up with. Is it possibly to make face quadrature order always depends on the volume quadrature order? This could be a painful design discussion.

rwcarlsen commented:
https://github.com/idaholab/moose/issues/15472
